tests/integration/utils.sh

```diff
@@ -157,6 +158,123 @@ function teardown_single_container()
+  local n="${#DISTROS[@]}" # total number of steps of tear down
+
+  for distro in "${DISTROS[@]}"; do
+    # progress message
+    i=$((i + 1))
+    say "[${i}/${n}] Removing ${distro} container."
+
+    teardown_single_container "kw-${distro}"
+  done
+}
+
+# Completely remove the container environment: the containers and the images
+function teardown_container_environment()
+{
+  local flags="$1"
+  local container_name
+  local distro
+  local img
+  local img_layers
+  local tmp_img
+  local tmp_img_layers
+  local tmp_output
+
+  for distro in "${DISTROS[@]}"; do
+    container_name="kw-${distro}"
+    podman container exists "${container_name}"
+    if [[ "$?" -eq 0 ]]; then
+      say "Removing podman container kw-${distro}."
+      teardown_single_container "${container_name}"
+    fi
+
+    img="kw-${distro}"
+    podman image exists "${img}"
+    if [[ "$?" -eq 0 ]]; then
+      say "Removing podman ${img} image."
+
+      # Get this image layers, so we can safely remove its children dangling images.
+      img_layers=$(image_inspect -f '{{.RootFS.Layers}}' "${img}" | tr -d '[]')
+
+      # Gemove possible dangling images
+      while read -r tmp_img; do
+        # We check if the image exist  because  it  could  happen  that  in  the
+        # previous iteration the father dangling iamge was deleted, causing  the
+        # child dangling image to be deleted as well.
+        podman image exists "${tmp_img}"
+        if [[ "$?" -ne 0 ]]; then
+          continue
+        fi
+
+        # We check if the dangling image have layers which are exact prefixes of
+        # the current image we are trying to delete.
+        tmp_img_layers=$(image_inspect -f '{{.RootFS.Layers}}' "${tmp_img}" | tr -d '[]')
+        tmp_output=$(grep --perl-regexp "^${tmp_img_layers}" <<< "${img_layers}")
+
+        # Delete the dangling image if that is the case.
+        if [[ -n "$tmp_output" ]]; then
+          # Dangling images are always force-removed.
+          image_rm --force "${tmp_img}"
+        fi
+      done <<< "$(image_ls --all --quiet --filter 'intermediate=true')"
+
+      # Remove main image.
+      image_rm "${flags}" "${img}"
+      if [[ "$?" -ne 0 ]]; then
+        complain "Failed to remove ${img} image."
+      fi
+    fi
+
+    # Remove the base image for the distro.
+    img="${CONTAINER_BASE_IMAGE}/${distro}"
+    podman image exists "${img}"
+    if [[ "$?" -eq 0 ]]; then
+      say "Removing podman ${img} image."
+      image_rm "${flags}" "${img}"
+      if [[ "$?" -ne 0 ]]; then
+        complain "Failed to remove ${img} image."
+      fi
+    fi
+  done
+}
+
+function image_rm()
+{
+  # shellcheck disable=SC2068
+  podman image rm $@ > /dev/null
+
+  if [[ "$?" -ne 0 ]]; then
+    complain "(${LINENO}): Failed to remove the images."
+  fi
+}
+
+function image_ls()
+{
+  # shellcheck disable=SC2068
+  podman image ls $@
+
+  if [[ "$?" -ne 0 ]]; then
+    complain "(${LINENO}): Failed to list the images."
```

Comment:
To make debugging easier in case of failure, add the command you tried to execute as part of the error message. Something like: `kw failed to execute the command: podman image ls SOMETHING SOMETHING`

Reply:
ok